File: RxTango/python/examples/running_stats.py
"""Live streaming statistics using ops.scan (Welford's online algorithm).

Tracks count, mean, and standard deviation without storing all past values.

Mirrors `TangoTestRunningStats.java`.

Usage:
    python running_stats.py [device] [interval-ms]

    device       defaults to tango://localhost:10000/sys/tg_test/1
    interval-ms  defaults to 500
"""

import asyncio
import math
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute


def welford_update(state, x):
    """Welford's online algorithm step: (n, mean, M2) → updated state."""
    n, mean, m2 = state
    n += 1
    delta = x - mean
    mean += delta / n
    m2 += delta * (x - mean)
    return n, mean, m2


async def main() -> None:
    device      = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    interval_ms = int(sys.argv[2]) if len(sys.argv) > 2 else 500

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    print(f"Running stats on {device}  every {interval_ms} ms  (Ctrl+C to stop)\n")
    print(f"  {'n':>6}  {'value':>14}  {'mean':>14}  {'std':>14}")
    print("  " + "-" * 56)

    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        ops.flat_map(lambda _: read_attribute(device, "double_scalar")),
        ops.scan(welford_update, seed=(0, 0.0, 0.0)),
    ).subscribe(
        on_next=lambda s: print(
            f"  {s[0]:>6}  {0.0:>14.6f}  {s[1]:>14.6f}  "
            f"{math.sqrt(s[2] / s[0]) if s[0] > 1 else 0.0:>14.6f}"
        ),
        on_error=lambda e: print(f"  ERROR: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n  stopped.")

File: RxTango/cpp/examples/running_stats.cpp
/**
 * Live O(1) running statistics using Welford's online algorithm via scan().
 *
 * No buffer needed — scan() accumulates (count, mean, M2) in constant space.
 * On each new value: count++, update mean and variance.
 *
 * Mirrors Python's running_stats.py and Java's TangoTestRunningStats.java.
 *
 * Usage:
 *   ./running_stats [device] [attribute] [interval-ms]
 *   defaults: sys/tg_test/1  double_scalar  200
 */

#include <chrono>
#include <csignal>
#include <cmath>
#include <iostream>
#include <string>
#include <tuple>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

// Welford accumulator: (count, mean, M2)
using Stats = std::tuple<int, double, double>;

static Stats welford_update(Stats acc, double x) {
    auto [n, mean, M2] = acc;
    ++n;
    double delta  = x - mean;
    mean         += delta / n;
    double delta2 = x - mean;
    M2           += delta * delta2;
    return {n, mean, M2};
}

int main(int argc, char* argv[]) {
    const std::string device      = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr        = argc > 2 ? argv[2] : "double_scalar";
    const int         interval_ms = argc > 3 ? std::stoi(argv[3]) : 200;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Running stats for " << device << "/" << attr << "  (Ctrl+C to stop)\n\n"
              << std::setw(8) << "N" << std::setw(14) << "mean"
              << std::setw(14) << "stddev" << "\n"
              << std::string(38, '-') << "\n";

    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([device, attr](long) {
            return rxtango::read_attribute<double>(device, attr);
        })
        .scan(Stats{0, 0.0, 0.0}, welford_update)
        .subscribe(
            [](Stats s) {
                auto [n, mean, M2] = s;
                double stddev = (n > 1) ? std::sqrt(M2 / (n - 1)) : 0.0;
                std::cout << std::setw(8) << n
                          << std::setw(14) << mean
                          << std::setw(14) << stddev << "\n";
            },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    return 0;
}
